control/control.rnc

```diff
@@ -566,9 +566,14 @@ system_role = element system_role {
     ## texts section: ${id} and ${id}_description.
     element id { text } &
 
-    ## option to not preselect any role, by default first one is preselected
+    ## option to not preselect any role, by default first one by order is preselected
     element no_default { BOOLEAN }? &
 
+    ## Option for order of role. Roles are sorted by order, the smaller first. It can be negative.
+    ## If not specified default 10 000 is used. If two roles have same order, then the first come
```

Comment:
The default number 10 000 is too high IMHO, I think 1 000 is enough. If you look at the pattern order numbers (`rpm -qa --provides | grep pattern-order | sort`) the numbers are not that high.

On my TW system it's from 1020 to 5190, but actually only 3 are above 3000, the vast majority is between 1 000 and 2 000. We should use or expect similar numbers here.

And I'm not sure we need negative numbers with that large range. Usually it should be always possible to use a
bit smaller number. It can be supported in the code but I'd not mention it here (and keep it as a secret hack for the future ;-))

Reply:
I use 10 000 to have roles without defined order as last. So I want it big enough.

With negative number, I just do't want to have limitation at zero as there is no real need for it and it allow to almost always find smaller number. ( especially in ruby with BigNum ).

Reply:
When the order is missing then it is simply `nil`. Then just add those cards at the end, no need for any value for them.

Just a note: BigNum will probably not work as the profile is read by the C++ XML agent which is limited to `YCPInteger` range (64 bit), but who would use such big numbers anyway...

Reply:
BTW there no ordering yet, so having the default 1 000 or 10 000 does not make any difference now.

Reply:
well, I also consider nil and having it as last, but this 

1. make unnecessary limit ( cannot specify in order that this is the last item, so even after ones without order )

2. makes code harder ( as nil is no number ).

Good point with C++ XML agent. Maybe if someone will use it and we finally switch to more reasonable xml parser like nokogiri

Reply:
1. All role definitions are under our control (at least now), we should add the numbers everywhere. Then adding the last item after undefined ones is rather a theoretical issue.
2. Yes, but that's trivial `nil?` filter and merging back after sorting, 2 lines more do not hurt

Reply:
1. do not forget that I do not due to card that adds ability to add roles to any module. So e.g. even SLE manager team can add it and won't maintain it.

Reply:
$2c: If a role does not have a priority (order), then it just goes to the end. To make it somehow "stable", then sorted alphabetically.

Reply:
I agree that using a "magic number" like 10 000 might come as some surprise to many users. While it is clearly documented here, people might still overlook this; also, you'll never know what others who make some product on top of SLE might use. They might think to play it safe and start their own roles from 10 000, so they might get their own roles 10 000, 10 100, 10 2000 etc., only to be surprised by roles that don't have any order defined to be sorted in between the original ones from the SLE base and their self-defined ones. Agreed, this is a somewhat contrived example, but just leaving it empty and using 'nil' for this sounds a lot more natural (and more Rubyish) and less error-prone to me.

Reply:
just to get impression about implementation with current default number ( I already have it prepared ).

```
@roles = raw_roles.map { |r| from_control(r) }
# do stable sorting, so if order is not defined, it keeps order
# for code reference see https://stackoverflow.com/questions/15442298/is-sort-in-ruby-stable
@roles.sort_by!.with_index { |role, index| [role.order, index] }
```

Reply:
ok, so I change it to nil.

Reply:
I do not expect the SUSE Manager team will want their role to be the very last one when you register the Manager extension... The usual question will be what will be on the top, not what will be the last item.

And most importantly, as the `order` element is required in your definition (missing `?` to make it optional). That means there should not be a role with undefined ordering. That means we do not need to define the default at all, we just should recommend the range to use.

And for the other ordering number we already use much smaller numbers (https://github.com/yast/skelcd-control-leanos/blob/master/control/control.leanos.xml#L230) I really do not see a reason for such big number.